src/librbd/cache/rwl/ImageCacheState.cc

```diff
@@ -83,6 +97,104 @@ void ImageCacheState<I>::dump(ceph::Formatter *f) const {
   ::encode_json("rwl_size", size, f);
 }
 
+template <typename I>
+void ImageCacheState<I>::get_image_cache_state(I *image_ctx,
+                                               Context *on_finish) {
+  std::string cache_state_str;
+  ldout(image_ctx->cct, 20) << "image_cache_state:" << cache_state_str << dendl;
+
+  bool dirty_cache = image_ctx->test_features(RBD_FEATURE_DIRTY_CACHE);
+  if (dirty_cache) {
+    cls_client::metadata_get(&image_ctx->md_ctx, image_ctx->header_oid,
+                             IMAGE_CACHE_STATE, &cache_state_str);
+  }
+
+  bool rwl_enabled = image_ctx->config.template get_val<bool>("rbd_rwl_enabled");
+  bool cache_desired = rwl_enabled;
+  cache_desired &= !image_ctx->read_only;
+  cache_desired &= !image_ctx->test_features(RBD_FEATURE_MIGRATING);
+  cache_desired &= !image_ctx->test_features(RBD_FEATURE_JOURNALING);
+  cache_desired &= !image_ctx->old_format;
+
+  if (!dirty_cache && !cache_desired) {
+    ldout(image_ctx->cct, 20) << "Do not desire to use image cache." << dendl;
+    image_ctx->cache_state = nullptr;
+    on_finish->complete(0);
+    return;
+  } else if (dirty_cache && !cache_desired) {
+    lderr(image_ctx->cct) << "There's a dirty cache, but RWL cache is disabled "
+                          << "in ceph.conf" << dendl;
```

Comment:
Nit: drop the reference to "ceph.conf" since it could also be disabled numerous other ways.

Reply:
DONE